pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,660 @@
+	}
+}
+
+func (c *leaseState) startElection(total int) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.candidatesAck = make(map[string]bool)
+	c.done = make(chan bool)
+	c.total = total
+}
+
+func (c *leaseState) getValidCandidates() []string {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.candidatesAck == nil {
+		return nil
+	}
+	var candidates []string
+	for c := range c.candidatesAck {
+		candidates = append(candidates, c)
+	}
+	return candidates
+}
+
+func (c *leaseState) endElection() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.done = nil
+	c.candidatesAck = nil
+	c.total = -1
+}
+
+func (c *Controller) getLS(leaderLeaseID leaderLeaseID) *leaseState {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	ct, ok := c.leaseState[leaderLeaseID]
+	if ok {
+		return ct
+	}
+	c.leaseState[leaderLeaseID] = &leaseState{}
+	return c.leaseState[leaderLeaseID]
+
+}
+
+type leaderLeaseID struct {
+	namespace, name string
+}
+
+func (l leaderLeaseID) String() string {
+	return l.namespace + "/" + l.name
+}
+
+func (c *Controller) Run(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+	defer c.leaseQueue.ShutDown()
+	defer c.electionQueue.ShutDown()
+
+	if !cache.WaitForNamedCacheSync(controllerName, ctx.Done(), c.leaseSynced, c.leaseCandidateSynced) {
+		return
+	}
+	klog.Infof("Workers: %d", workers)
+	for i := 0; i < workers; i++ {
+		klog.Infof("Starting worker")
+		go wait.UntilWithContext(ctx, c.runWorker, time.Second)
+		go wait.UntilWithContext(ctx, c.runElectionWorker, time.Second)
+	}
+	<-ctx.Done()
+}
+
+func NewGenerator(ctx context.Context, config *rest.Config, leaseInformer coordinationv1informers.LeaseInformer, leaseCandidateInformer coordinationv1alpha1.LeaseCandidateInformer, leaseClient coordinationv1client.CoordinationV1Interface, leaseCandidateClient coordinationv1alpha1client.CoordinationV1alpha1Interface) (*Controller, error) {
+	var cancel context.CancelFunc
+
+	callbacks := leaderelection.LeaderCallbacks{
+		OnStartedLeading: func(ctx context.Context) {
+			var genCtx context.Context
+			controller, err := NewController(leaseInformer, leaseCandidateInformer, leaseClient, leaseCandidateClient)
+			if err != nil {
+				panic(err)
+			}
+			genCtx, cancel = context.WithCancel(ctx)
+			controller.Run(genCtx, 2)
+		},
+		OnStoppedLeading: func() {
+			cancel()
+		},
+	}
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+
+	rl, err := resourcelock.NewFromKubeconfig(
+		"leases",
+		"kube-system",
+		controllerName,
+		resourcelock.ResourceLockConfig{
+			Identity: hostname + "_" + string(uuid.NewUUID()),
+		},
+		config,
+		10,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	le, err := leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: 10 * time.Second,
+		RenewDeadline: 5 * time.Second,
+		RetryPeriod:   2 * time.Second,
+		Callbacks:     callbacks,
+		Name:          controllerName,
+	})
+	if err != nil {
+		return nil, err
+	}
+	go le.Run(ctx)
```

Comment:
go routines inside of a constructor are forbidden. Move to some Run or Start method.